chevah/keycert/tests/test_ssl.py

```diff
@@ -0,0 +1,395 @@
+            ])
+
+        with self.assertRaises(KeyCertException) as context:
+            generate_csr(options)
+
+        self.assertEqual('Unknown key type.', context.exception.message)
+
+    def test_bad_size(self):
+        """
+        Raise an exception when failing to generate the key.
+        """
+        options = self.parseArguments([
+            self.command_name,
+            '--common-name=domain.com',
+            '--key-type=rsa',
+            '--key-size=12',
+            ])
+
+        with self.assertRaises(KeyCertException) as context:
+            generate_csr(options)
+
+        self.assertEqual('key size too small', context.exception.message)
+
+    def test_bad_county_long(self):
+        """
+        Raise an exception when country code is not correct.
+        """
+        options = self.parseArguments([
+            self.command_name,
+            '--common-name=domain.com',
+            '--country=USA',
+            ])
+
+        with self.assertRaises(KeyCertException) as context:
+            generate_csr(options)
+
+        self.assertEqual('string too long', context.exception.message)
+
+    def test_bad_county_short(self):
+        """
+        Raise an exception when country code is not correct.
+        """
+        options = self.parseArguments([
+            self.command_name,
+            '--common-name=domain.com',
+            '--country=A',
+            ])
+
+        with self.assertRaises(KeyCertException) as context:
+            generate_csr(options)
+
+        self.assertEqual('string too short', context.exception.message)
+
+    def test_default_gen(self):
+        """
+        By default it will serialized the key without password and generate
+        the csr without alternative name and just the common name.
+        """
+        options = self.parseArguments([
+            self.command_name,
+            '--common-name=domain.com',
+            ])
+
+        result = generate_csr(options)
+
+        # OpenSSL.crypto.PKey has no equality so we need to compare the
+        # serialization.
+        self.assertEqual(2048L, result['key'].bits())
+        self.assertEqual(crypto.TYPE_RSA, result['key'].type())
+        key = crypto.dump_privatekey(crypto.FILETYPE_PEM, result['key'])
+        self.assertEqual(key, result['key_pem'])
+        # For CSR we can not get extensions so we only check the subject.
+        csr = crypto.dump_certificate_request(
+            crypto.FILETYPE_PEM, result['csr'])
+        self.assertEqual(csr, result['csr_pem'])
+        subject = result['csr'].get_subject()
+        self.assertEqual(u'domain.com', subject.commonName)
+        self.assertIsNone(subject.emailAddress)
+        self.assertIsNone(subject.organizationName)
+        self.assertIsNone(subject.organizationalUnitName)
+        self.assertIsNone(subject.localityName)
+        self.assertIsNone(subject.stateOrProvinceName)
+        self.assertIsNone(subject.countryName)
```

Comment:
alternative name is installed as an extension and the current API does not allow getting an extension... I know is strange... pyopenssl 0.15 will add this option.. but is not yet released